### ISSUE TITLE
how to deal with data with multiple cells

### ISSUE THREAD

**Issue by ch8316f5eyu** [PacificBiosciences/IsoSeq3#10]
If I have multiple cells, can I merge them together to do the iso-seq?
For example, I use "Sequel System Data Release: Iso-Seq Results for Hummingbird and Zebra Finch Brain Tissue" to perform iso-seq3.
If I perform iso-seq3 per cell, the result is OK. But when I merge 4 subreads.bam firstly, the result is wrong. The polished.hq.fasta.gz file has nothing. I think the problem is in the "isoseq3 polish" step.
I think using the merged bam file to perform downstream analysis will get more hq sequence.
 So, what is the correct way to deal with data with multiple cells?

**Comment by Magdoll** [comment]
Hi @ch8316f5eyu ,
Please follow the examples [in this tutorial](https://github.com/PacificBiosciences/IsoSeq_SA3nUP/wiki/Iso-Seq:-misc-knowledge) about merging subreads.bam files.

You will need to create a `.subreadset.xml` file and use that for the polishing step.

--Liz

**Comment by ch8316f5eyu** [comment]
Thank you for your reply.
But I cannot find the dataset create command. Is it the dataset.py form PacificBiosciences/pbcore? After I finished installing pbcore, I cannot find the dataset.py script.

**Comment by Magdoll** [comment]
Hi @ch8316f5eyu ,

`dataset` command is part of [pbcoretools](https://github.com/PacificBiosciences/pbcoretools), which is not on bioconda yet. However at this point you can install it using pip:

```
pip install https://github.com/PacificBiosciences/pbcoretools/archive/master.zip
```

**Comment by mhsieh** [comment]
Alternatively (I prefer this way better :p ) the dataset command can be from our smrtlink software distribution. You can download the zip file from our website, unzip with the password provide on the website, and run
```
bash smrtlink_5.1.0.26412.run --rootdir <ThePathWhereYouWantToExtract> --smrttools-only
```
Suppose that you extracted it to `/opt/pacbio/smrttools`, you can then use the dataset command as `/opt/pacbio/smrttools/smrtcmds/bin/dataset`

**Comment by armintoepfer** [comment]
I agree with MJ. Use smrttools and you'll have everything you can't get through bioconda yet.

**Comment by armintoepfer** [comment]
Did you generate pbi files for each subreads.bam?

https://github.com/PacificBiosciences/IsoSeq3/issues/6

**Comment by ch8316f5eyu** [comment]
Yes. I download everything from RawData directory from https://downloads.pacbcloud.com/public/dataset/AvianBrainTranscriptome/.

**Comment by armintoepfer** [comment]
Someone from our team will have a look at it.

**Comment by Magdoll** [comment]
Hi @ch8316f5eyu ,

Did you say you used samtools to merge the input subreads from [RawData dir](https://downloads.pacbcloud.com/public/dataset/AvianBrainTranscriptome/)?  How did you generate the file `/disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/All.xml` then?

I think you did not have to merge the subreads.bam files and you only needed to provide a correct dataset XML. It should be like:

```
dataset create --type SubreadSet --name All.subreadset.xml \
     <full_path_to>/m54006_170729_232022.subreads.bam \
     <full_path_to>/m54026_170727_103805.subreads.bam  \
     <full_path_to>/m54200_170721_210832.subreads.bam  \
     <full_path_to>/m54200_170722_173443.subreads.bam 
```

Then run polish with `All.subreadset.xml` provided. Please let me know if this works. The error message points the problem towards the dataset XML.

--Liz

**Comment by ch8316f5eyu** [comment]
Sorry, I am a little confused. I did use samtools to merge input subreads first, and did ccs, lima, cluster. And I provide all.xml file to polish as you said, then the error message occurred.
Do you mean that I shouldn't merge the input subreads at first?
Then, what is the correct workflow?

**Comment by ch8316f5eyu** [comment]
@Magdoll 
This time I did not use samtools to merge bam file, but the error still occurred.
This is my code:
`dataset create --type SubreadSet /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/All.xml /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54006_170729_232022.subreads.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54026_170727_103805.subreads.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54200_170721_210832.subreads.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54200_170722_173443.subreads.bam

~/bin/bin/ccs /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/All.xml /disk5/Data/Pacbio/AvianBrainTranscriptome/Ccsdata/All.ccs.bam --noPolish --numThreads 20 --minPasses 1

lima /disk5/Data/Pacbio/AvianBrainTranscriptome/Ccsdata/All.ccs.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/barcode.fa /disk5/Data/Pacbio/AvianBrainTranscriptome/Demux/All.ccs.demux.bam --isoseq --no-pbi --split-bam-named --num-threads 20

isoseq3 cluster --num-threads 20 /disk5/Data/Pacbio/AvianBrainTranscriptome/Demux/All.ccs.demux.primer_5p--hummingbird_3p.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Cluster/hummingbird.unpolished.bam
isoseq3 cluster --num-threads 20 /disk5/Data/Pacbio/AvianBrainTranscriptome/Demux/All.ccs.demux.primer_5p--Zebra_finch_3p.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Cluster/Zebra_finch.unpolished.bam

isoseq3 polish /disk5/Data/Pacbio/AvianBrainTranscriptome/Cluster/hummingbird.unpolished.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/All.xml /disk5/Data/Pacbio/AvianBrainTranscriptome/Polish/hummingbird.polished.bam
isoseq3 polish /disk5/Data/Pacbio/AvianBrainTranscriptome/Cluster/Zebra_finch.unpolished.bam /disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/All.xml /disk5/Data/Pacbio/AvianBrainTranscriptome/Polish/Zebra_finch.polished.bam
`

**Comment by Magdoll** [comment]
Hi @ch8316f5eyu ,
You you confirm that the .pbi index exists for the bam files? 

They would be

```
/disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54006_170729_232022.subreads.bam.pbi
/disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54026_170727_103805.subreads.bam.pbi
/disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54200_170721_210832.subreads.bam.pbi
/disk5/Data/Pacbio/AvianBrainTranscriptome/Rawdata/m54200_170722_173443.subreads.bam.pbi
```

currently the polish step will silently fail if the `.pbi` files don't exist. You can create the index using `pbindex`.

If you are still failing the `isoseq3 polish` step, maybe there's some subtle error we are not realizing. In that case please give me an email and I will ask you to upload `/disk5/Data/Pacbio/AvianBrainTranscriptome/Cluster/hummingbird.unpolished.bam ` to test it ourselves.

--Liz

**Comment by fantastycrane** [comment]
Hi @Magdoll ,

It says in the workfow that execute subsequent steps for each output BAM file after lima. I notice that ch8316f5eyu merged data from all cells before ccs. Can we perform ccs and lima on each cell separately, then create a merged .xml of all BAM files generated by lima and run cluster?

**Comment by ch8316f5eyu** [comment]
The error occurred because of incompleted pbi file. One pbi file is incomplete. It is my fault. In addition to this, the polish step generated a lot of information, for example, 
`>|> 20180726 05:18:27.065 -|- INFO       -|- Polish -|- 0x7fce197ea700|| -|- negative inf in mutation testing: 'm54006_170729_232022/54854172/8826_19252'`.
What does it mean?
